doc/designs/functions-for-explicitly-fetched-signature-algorithms.md

```diff
@@ -0,0 +1,181 @@
+Functions for explicitly fetched PKEY algorithms
+================================================
+
+Quick background
+----------------
+
+There are several proposed designs that end up revolving around the same
+basic need, explicitly fetched signature algorithms.  The following method
+type is affected by this document:
+
+- `EVP_SIGNATURE`
+
+Public API - Add variants of `EVP_PKEY_CTX` functionality
+---------------------------------------------------------
+
+Through OTC discussions, it's been determined that the most suitable APIs to
+touch is the collection of `EVP_PKEY_` functions.  In this case, it involves
+`EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()` with
+associated functions.  They can be extended to accept an explicitly fetched
+algorithm of the right type, and to be able to process infinite amounts of
+data if the fetched algorithm permits it (typically, algorithms like ED25519
+or RSA-SHA256).
+
+It must be made clear that the added functionality can not be used to
+compose an algorithm from different parts.  For example, it's not possible
+to specify a `EVP_SIGNATURE` "RSA" and combine it with a parameter that
+specifies the hash "SHA256" to get the "RSA-SHA256" functionality.  For an
+`EVP_SIGNATURE` "RSA", the input is still expected to be a digest, or some
+other input that's limited to the modulus size of the RSA pkey.
+
+### Making things less confusing with distinct function names
+
+So far, `EVP_PKEY_sign()` and friends are only expected to act as
+"primitive" functions.  This design essentially proposes an extension to
+also allow streaming functionality through an *update* and a *final*
+function.  Discussions have revealed that it is potentially confusing to
+confound "primitive" functionality with "streaming" functionality into the
+same name, so this design also includes a naming proposal that distinguishes
+the two uses.
+
+### Making it possible to verify with an early signature
+
+There are newer verifying algorithms that need to receive the signature
+before processing the data.  This is particularly important for streaming
+functionality.  This design proposes a mechanism to accomodate this, which
+is different than the mechanism you get with `EVP_PKEY_verify()`.
+
+Public API - API Reference
+--------------------------
+
+### For signing with `EVP_SIGNATURE`
+
+#### Primitive / Limit size oneshot
+
+``` C
+int EVP_PKEY_sign_primitive_init(EVP_PKEY_CTX *pctx,
+                                 EVP_SIGNATURE *algo,
+                                 const OSSL_PARAM params[]);
+int EVP_PKEY_sign_primitive(EVP_PKEY_CTX *ctx,
+                            unsigned char *sig, size_t *siglen, size_t sigsize,
+                            const unsigned char *tbs, size_t tbslen);
```

Comment:
I assumed this will be just an alias to EVP_PKEY_sign().

Reply:
Implementation detail. But yeah